lib/maths/CBayesianOptimisation.cc

```diff
@@ -189,6 +205,166 @@ CBayesianOptimisation::maximumExpectedImprovement() {
+    TVector scaledKernelParameters{this->scaledKernelParameters()};
+    auto prodXt = [&scaledKernelParameters, this](const TVector& x, int t) {
+        double prod{1.0};
+        for (int d = 0; d < m_MinBoundary.size(); ++d) {
+            if (d != t) {
+                prod *= stableProdd(scaledKernelParameters(d), x(d));
+            }
+        }
+        return prod;
+    };
+    double f0{this->anovaConstantFactor(Kinvf)};
+    double sum{0.0};
+    input = (input - m_MinBoundary(dimension)) /
+            (m_MaxBoundary(dimension) - m_MinBoundary(dimension));
+    for (std::size_t i = 0; i < m_FunctionMeanValues.size(); ++i) {
+        TVector x{this->transformTo01(m_FunctionMeanValues[i].first)};
+        sum += Kinvf(static_cast<int>(i)) *
+               std::exp(-(CTools::pow2(scaledKernelParameters[dimension]) +
+                          MINIMUM_KERNEL_COORDINATE_DISTANCE_SCALE) *
+                        CTools::pow2(input - x(dimension))) *
+               prodXt(x, dimension);
+    }
+    return CTools::pow2(m_KernelParameters(0)) * sum - f0;
+}
+
+double CBayesianOptimisation::evaluate1D(double input, int dimension) const {
+    return this->evaluate1D(this->kinvf(), input, dimension);
+}
+
+double CBayesianOptimisation::anovaConstantFactor(const TVector& Kinvf) const {
+    TVector scaledKernelParameters{this->scaledKernelParameters()};
+    double sum{0.0};
+    for (std::size_t i = 0; i < m_FunctionMeanValues.size(); ++i) {
+        double prod{1.0};
+        TVector x{this->transformTo01(m_FunctionMeanValues[i].first)};
+        for (int d = 0; d < m_MinBoundary.size(); ++d) {
+            prod *= stableProdd(scaledKernelParameters(d), x(d));
+        }
+        sum += Kinvf(i) * prod;
+    }
+    return CTools::pow2(m_KernelParameters(0)) * sum;
+}
+
+double CBayesianOptimisation::anovaConstantFactor() const {
+    return this->anovaConstantFactor(this->kinvf());
+}
+
+double CBayesianOptimisation::anovaTotalVariance(const TVector& Kinvf) const {
+    TVector scaledKernelParameters{this->scaledKernelParameters()};
+    auto prodIj = [this, &scaledKernelParameters, &Kinvf](std::size_t i, std::size_t j) {
+        TVector xi{this->transformTo01(this->m_FunctionMeanValues[i].first)};
+        TVector xj{this->transformTo01(this->m_FunctionMeanValues[j].first)};
+        double prod{1.0};
+        for (int t = 0; t < this->m_MinBoundary.size(); ++t) {
+            prod *= stableProdIjt(scaledKernelParameters[t], xi(t), xj(t));
+        }
+        return Kinvf(static_cast<int>(i)) * Kinvf(static_cast<int>(j)) * prod;
+    };
+
+    double sum{0.0};
+    for (std::size_t i = 0; i < m_FunctionMeanValues.size(); ++i) {
+        for (std::size_t j = 0; j < m_FunctionMeanValues.size(); ++j) {
+            sum += prodIj(i, j);
+        }
+    }
+    double variance{std::pow(m_KernelParameters(0), 4) * sum -
+                    CTools::pow2(this->anovaConstantFactor(Kinvf))};
```

Comment:
This will have rather bad cancellation errors if `m_KernelParameters(0)` is large and we know that the constant factor is proportional to `m_KernelParameters(0)^2`. Let's discuss numerics a bit offline.

Reply:
Also, I think we should be bounding variances to >= 0.0.

Reply:
I bounded the variance.